File: sprout/lexer.py
from __future__ import annotations

from .token import Token, TokenType, lookup_ident


class Lexer:
    def __init__(self, input_text: str) -> None:
        self.input_text = input_text
        self.position = 0
        self.read_position = 0
        self.ch = ""
        self._read_char()

    def _read_char(self) -> None:
        if self.read_position >= len(self.input_text):
            self.ch = "\0"
        else:
            self.ch = self.input_text[self.read_position]
        self.position = self.read_position
        self.read_position += 1

    def _peek_char(self) -> str:
        if self.read_position >= len(self.input_text):
            return "\0"
        return self.input_text[self.read_position]

    def _skip_whitespace(self) -> None:
        while self.ch in " \t\n\r":
            self._read_char()

    def _read_identifier(self) -> str:
        start = self.position
        while self.ch.isalpha() or self.ch == "_" or self.ch.isdigit():
            self._read_char()
        return self.input_text[start:self.position]

    def _read_number(self) -> str:
        start = self.position
        while self.ch.isdigit():
            self._read_char()
        return self.input_text[start:self.position]

    def _read_string(self) -> str:
        # current ch is opening quote
        self._read_char()
        start = self.position
        while self.ch != '"' and self.ch != "\0":
            if self.ch == "\\" and self._peek_char() in ['"', "\\", "n", "t", "r"]:
                # consume escape and next char as literal pair
                self._read_char()
            self._read_char()
        literal = self.input_text[start:self.position]
        # closing quote
        self._read_char()
        # unescape common sequences
        return bytes(literal, "utf-8").decode("unicode_escape")

    def next_token(self) -> Token:
        self._skip_whitespace()

        if self.ch == "=":
            if self._peek_char() == "=":
                ch = self.ch
                self._read_char()
                tok = Token(TokenType.EQ, ch + self.ch, self.position)
                self._read_char()
                return tok
            tok = Token(TokenType.ASSIGN, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == ";":
            tok = Token(TokenType.SEMICOLON, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == ",":
            tok = Token(TokenType.COMMA, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == ":":
            tok = Token(TokenType.COLON, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "+":
            tok = Token(TokenType.PLUS, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "-":
            tok = Token(TokenType.MINUS, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "!":
            if self._peek_char() == "=":
                ch = self.ch
                self._read_char()
                tok = Token(TokenType.NOT_EQ, ch + self.ch, self.position)
                self._read_char()
                return tok
            tok = Token(TokenType.BANG, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "/":
            tok = Token(TokenType.SLASH, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "*":
            tok = Token(TokenType.ASTERISK, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "<":
            tok = Token(TokenType.LT, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == ">":
            tok = Token(TokenType.GT, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "(":
            tok = Token(TokenType.LPAREN, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == ")":
            tok = Token(TokenType.RPAREN, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "{" :
            tok = Token(TokenType.LBRACE, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "}":
            tok = Token(TokenType.RBRACE, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "[":
            tok = Token(TokenType.LBRACKET, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == "]":
            tok = Token(TokenType.RBRACKET, self.ch, self.position)
            self._read_char()
            return tok
        elif self.ch == '"':
            literal = self._read_string()
            return Token(TokenType.STRING, literal, self.position)
        elif self.ch == "\0":
            return Token(TokenType.EOF, "", self.position)
        else:
            if self.ch.isalpha() or self.ch == "_":
                ident = self._read_identifier()
                return Token(lookup_ident(ident), ident, self.position)
            if self.ch.isdigit():
                num = self._read_number()
                return Token(TokenType.INT, num, self.position)
            tok = Token(TokenType.ILLEGAL, self.ch, self.position)
            self._read_char()
            return tok

File: sprout-js/src/lexer.js
'use strict';

const { TokenType, lookupIdent } = require('./token');

class Lexer {
  constructor(input) {
    this.input = input;
    this.position = 0;
    this.readPosition = 0;
    this.ch = '';
    this.line = 1;
    this.column = 0;
    this._readChar();
  }

  _readChar() {
    if (this.readPosition >= this.input.length) {
      this.ch = '\0';
    } else {
      this.ch = this.input[this.readPosition];
      if (this.ch === '\n') {
        this.line++;
        this.column = 0;
      } else {
        this.column++;
      }
    }
    this.position = this.readPosition;
    this.readPosition += 1;
  }

  _peekChar() {
    if (this.readPosition >= this.input.length) return '\0';
    return this.input[this.readPosition];
  }

  _skipWhitespace() {
    while (true) {
      // Handle line comments
      if (this.ch === '/' && this._peekChar() === '/') {
        while (this.ch !== '\n' && this.ch !== '\0') {
          this._readChar();
        }
        // Don't continue here, let the newline be processed normally
      }
      
      if (this.ch === ' ' || this.ch === '\t' || this.ch === '\r' || this.ch === '\n') {
        this._readChar();
        continue;
      }
      break;
    }
  }

  _readIdentifier() {
    const start = this.position;
    while (/[a-zA-Z_0-9]/.test(this.ch)) {
      this._readChar();
    }
    return this.input.slice(start, this.position);
  }

  _readNumber() {
    const start = this.position;
    let hasDecimal = false;
    
    // Read integer part
    while (/[0-9]/.test(this.ch)) {
      this._readChar();
    }
    
    // Check for decimal point
    if (this.ch === '.' && /[0-9]/.test(this._peekChar())) {
      hasDecimal = true;
      this._readChar(); // consume the decimal point
      
      // Read decimal part
      while (/[0-9]/.test(this.ch)) {
        this._readChar();
      }
    }
    
    const numStr = this.input.slice(start, this.position);
    return hasDecimal ? parseFloat(numStr) : parseInt(numStr, 10);
  }

  _readString() {
    this._readChar(); // consume opening quote
    const start = this.position;
    
    while (this.ch !== '"' && this.ch !== '\0') {
      if (this.ch === '\\') {
        const nextChar = this._peekChar();
        if (['"', '\\', 'n', 't', 'r'].includes(nextChar)) {
          this._readChar(); // consume the escape character
        }
      }
      this._readChar();
    }
    
    if (this.ch === '\0') {
      throw new Error(`Unterminated string at line ${this.line}, column ${this.column}`);
    }
    
    const raw = this.input.slice(start, this.position);
    this._readChar(); // consume closing quote
    
    // Decode simple escapes using JSON parse trick
    try {
      return JSON.parse('"' + raw.replace(/"/g, '\\"') + '"');
    } catch (e) {
      return raw;
    }
  }

  nextToken() {
    this._skipWhitespace();

    let tok;
    switch (this.ch) {
      case '=':
        if (this._peekChar() === '=') {
          const ch = this.ch; 
          this._readChar();
          tok = { type: TokenType.EQ, literal: ch + this.ch, line: this.line, column: this.column - 1 };
          this._readChar();
          return tok;
        }
        tok = { type: TokenType.ASSIGN, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case ';': 
        tok = { type: TokenType.SEMICOLON, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case ',': 
        tok = { type: TokenType.COMMA, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case ':': 
        tok = { type: TokenType.COLON, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '+': 
        tok = { type: TokenType.PLUS, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '-': 
        tok = { type: TokenType.MINUS, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '!':
        if (this._peekChar() === '=') {
          const ch = this.ch; 
          this._readChar();
          tok = { type: TokenType.NOT_EQ, literal: ch + this.ch, line: this.line, column: this.column - 1 };
          this._readChar();
          return tok;
        }
        tok = { type: TokenType.BANG, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '/':
        tok = { type: TokenType.SLASH, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '*': 
        tok = { type: TokenType.ASTERISK, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '<': 
        tok = { type: TokenType.LT, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '>': 
        tok = { type: TokenType.GT, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '(': 
        tok = { type: TokenType.LPAREN, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case ')': 
        tok = { type: TokenType.RPAREN, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '{': 
        tok = { type: TokenType.LBRACE, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '}': 
        tok = { type: TokenType.RBRACE, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '[': 
        tok = { type: TokenType.LBRACKET, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case ']': 
        tok = { type: TokenType.RBRACKET, literal: this.ch, line: this.line, column: this.column }; 
        this._readChar(); 
        return tok;
      case '"':
        const strValue = this._readString();
        return { type: TokenType.STRING, literal: strValue, line: this.line, column: this.column - 1 };
      case '\0':
        return { type: TokenType.EOF, literal: '', line: this.line, column: this.column };
      default:
        if (/[a-zA-Z_]/.test(this.ch)) {
          const ident = this._readIdentifier();
          return { 
            type: lookupIdent(ident), 
            literal: ident, 
            line: this.line, 
            column: this.column - ident.length + 1 
          };
        }
        if (/[0-9]/.test(this.ch)) {
          const num = this._readNumber();
          const isFloat = typeof num === 'number' && num % 1 !== 0;
          return { 
            type: isFloat ? TokenType.FLOAT : TokenType.INT, 
            literal: num, 
            line: this.line, 
            column: this.column - String(num).length + 1 
          };
        }
        tok = { 
          type: TokenType.ILLEGAL, 
          literal: this.ch, 
          line: this.line, 
          column: this.column 
        };
        this._readChar();
        return tok;
    }
  }
}

module.exports = { Lexer };

